### PR TITLE
🗞️ Native OFT Adapter

### PR DESCRIPTION
**Pull request by ravinagill15**
## In this PR
- Implemented `NativeOFTAdapter` with tests

### CODE REVIEW

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:17]: debit failed

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:19]: credit failed

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:51]: This doesnt do anything, the msg.value automatically moves all the funds to the contract

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:74]: function send(
    SendParam calldata _sendParam,
    MessagingFee calldata _fee,
    address _refundAddress
) external payable virtual override returns (MessagingReceipt memory msgReceipt, OFTReceipt memory oftReceipt) {
    if (_fee.nativeFee + _sendParam.amountLD != msg.value) {
        revert NotEnoughNative(msg.value);
    }
    
    return super.send(_sendParam, _fee, _refundAddress);
}

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:110]: override this to just do nothing

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:64]: // @dev  Transfer tokens to the recipient.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:86]: I would also comment the reason for muting this out was that the assertion is now done on the send higher up in this contract

**Comment by TRileySchwarz** [comment]
Also, need to add the token() and approvalRequired() functions

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:75]: @TRileySchwarz Do we still want to remove dust from the `amountLD` before sending? If so, I don't think we can remove dust from the `amountLD` here since `_sendParam` is of type `calldata`. Creating a new `sendParam` where we set `amountLD` with dust removed also won't work because it would be in memory, which doesn't align with the function parameters we end up passing it to

Or perhaps you meant the dust should already be removed in the incoming `_sendparam`?

Comment (TRileySchwarz) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:75]: we could in the assertion, mutate that param, but not adjust it in the super.send call. Alternatively I think we could just not do this and accumulate the dust

Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:75]: > we could in the assertion, mutate that param, but not adjust it in the super.send call

I don't think we can mutate it since it's of type `calldata`, but I suppose we could remove the dust in the assertion:
```
if (_fee.nativeFee + _removeDust(_sendParam.amountLD) != msg.value) {
            revert NotEnoughNative(msg.value);
}
```

Alternatively, If we accumulate the dust, i'm guessing we need some kind of withdraw function for it?

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:82]: Found out while writing tests that `msg.value` must be the same as `_fee.nativeFee`, even though we are overriding `_payNative` to be empty. This is due to `EndpointV2Mock._assertMessagingFee` requiring `msg.value` to be equal to `_fee.nativeFee` [here](https://github.com/LayerZero-Labs/devtools/blob/c745cf8d0992e2f8df7d00388596fe1b9c4b160b/packages/test-devtools-evm-foundry/contracts/mocks/EndpointV2Mock.sol#L318-L323)

I will look for a way around this

Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:82]: This actually may be due to the `msg.value` not being carried over to `EndpointV2Mock._assertMessagingFee`, still looking!

Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:82]: Figured this out -- `_payNative` cannot be empty, it must just return the parameter

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:75]: yeah sorry thats what i meant, do it like that. Technically does removeDust twice, but we can make this decision later

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:72]: comment this in detail

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:93]: summarize these comments better

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:120]: "...done higher up on the overriden send() function."

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:72]: comment this in detail, and add a second TODO comment about whether we want to keep the removeDust

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:8]: I would drop `Contract` here.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:23]: I would get rid of this line.  Don't need the @dev comment here as it doesn't add anything.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:17]: I will defer to @TRileySchwarz here, but we pretty much adapted the non-standard convention of using `address(0)` instead of this.  I am not quite sure what we want to support going forward, but historically we used `address(0)` as it is chain agnostic.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:44]: This documentation is off.  In this case, you are not using named return values.

File: packages/oft-evm/contracts/OFTCore.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/OFTCore.sol:177]: Why was this made public?

File: packages/oft-evm/test/OFT.t.sol
Comment (ryandgoulding) [inline packages/oft-evm/test/OFT.t.sol:66]: Add a comment like:

```
// must match NativeOFTAdapter.sol
```

**Review by ryandgoulding — COMMENTED** [review]
Great work!  A few small comments here and there, but structurally looks great.

File: packages/oft-evm/contracts/OFTCore.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/OFTCore.sol:177]: In order to call `super.send(...)` in `NativeOFTAdapter.send(...)` we had to change this from external to public. Do you foresee any issues with this change?

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:17]: Ah I see, makes sense! @TRileySchwarz I will change this to return `address(0)` based on @ryandgoulding's recommendation above, please lmk if you disagree

File: packages/oft-evm/test/OFT.t.sol
Comment (ravinagill15) [reply packages/oft-evm/test/OFT.t.sol:66]: I will be removing this in favour of using `address(0)` as [discussed above](https://github.com/LayerZero-Labs/devtools/pull/838#discussion_r1716762372)

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:71]: TODO has been added -- @TRileySchwarz @ryandgoulding what are your thoughts on whether we should remove dust from `amountLD` in the assertion here?

I think it makes sense to keep because it ensures the exact amount of `msg.value` is sent with no excess

File: .changeset/shaggy-actors-call.md
Comment (ravinagill15) [inline .changeset/shaggy-actors-call.md:2]: @TRileySchwarz @ryandgoulding @janjakubnanista Just want to confirm that this change would be considered minor?

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:72]: Wondering if instead of reverting with `NotEnoughNative` we should revert with a custom error that also indicates how much native _should_ have been provided (i.e `_fee.nativeFee + _removeDust(_sendParam.amountLD)`)

File: .changeset/shaggy-actors-call.md
Comment (ravinagill15) [reply .changeset/shaggy-actors-call.md:2]: Confirmed with @janjakubnanista that yes it would be a minor change

**Comment by ravinagill15** [comment]
Looking into build failures!

**Comment by ravinagill15** [comment]
Please do not merge until the NativeOFTAdapter contract has been audited

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:74]: it technically is the same purpose, but I could be convinced to do a separate error. The issue is that the original assertion for the fee->msg.value doesn't happen now, so the underlying behaviour is inconsistent

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (TRileySchwarz) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:73]: Calculate the extra gas for doing this, that will determine the outcome imo

File: packages/oft-evm/contracts/OFTCore.sol
Comment (TRileySchwarz) [reply packages/oft-evm/contracts/OFTCore.sol:177]: This wont effect things down the line, or break interfaces etc. we should probably make this public imo

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:73]: We only use an extra 194 gas when removing dust here vs without removing dust. Im thinking we should still keep it since its quite low?

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:74]: Gotcha, yeah in that case I'll add a new custom error specifically for this use case. It will also help the user see what they should have passed for `msgValue` if they accidentally pass an incorrect value

Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:74]: Actually, I think we should reuse `InsufficientMessageValue` error here as that already lets us indicate `provided` and `required` and fits this use case

File: packages/oft-evm/contracts/OFTCore.sol
Comment (ryandgoulding) [reply packages/oft-evm/contracts/OFTCore.sol:177]: It does make it slightly more expensive, but I see that it is needed.  I don't disagree.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (janjakubnanista) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:115]: @TRileySchwarz not sure whether it's valuable to pass the `data` from the `call` to the error for debugging, whatcha think?

File: packages/oft-evm/contracts/OFTCore.sol
Comment (janjakubnanista) [reply packages/oft-evm/contracts/OFTCore.sol:177]: What about moving the body to `internal` function (`_send` or smth) and using that instead of modifying the interface?

File: packages/oft-evm/test/OFT.t.sol
Comment (janjakubnanista) [inline packages/oft-evm/test/OFT.t.sol:457]: Two things:

- I would add comments about what's being tested and why it should fail
- I would add the missing case for when the `msgValue` is higher than `fee.nativeFee + sendParam.amountLD` because that triggers a revert as well

File: packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol
Comment (janjakubnanista) [inline packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol:26]: In general (so maybe in this case it's not applicable) testing internal functions is not the best idea. Internal functions can change as long as the external interface does not.

In these cases, I recommend testing these in the context of the external flow as they will never be run in isolation.

Check with @TRileySchwarz and @ryandgoulding though before changing anything, maybe there is some value to having them tested.

File: packages/oft-evm/contracts/OFTCore.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/OFTCore.sol:177]: I like that idea because it creates a consistent pattern between `OFTCore.send` and `NativeOFTAdapter.send` that can be referenced in similar future implementations.

I have a PR here that does exactly that: https://github.com/LayerZero-Labs/devtools/pull/877

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:115]: That sounds like a good idea to me! I have a follow up PR here: https://github.com/LayerZero-Labs/devtools/pull/878

File: packages/oft-evm/test/OFT.t.sol
Comment (ravinagill15) [reply packages/oft-evm/test/OFT.t.sol:457]: > I would add comments about what's being tested and why it should fail

Will do!

> I would add the missing case for when the msgValue is higher than fee.nativeFee + sendParam.amountLD because that triggers a revert as well

Oo yes great catch! I'll write this in the next commit

File: packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol
Comment (ravinagill15) [reply packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol:26]: I see what you mean, makes sense! I had originally done this to follow the [testing pattern of OFTAdapter](https://github.com/LayerZero-Labs/devtools/blob/555752ef24c787164cecd551aad363978da4d381/packages/oft-evm/test/OFT.t.sol#L329-L374), but if @TRileySchwarz and @ryandgoulding also agree, then I have a PR [here](https://github.com/LayerZero-Labs/devtools/pull/879) that removes testing internal functions. We already test the external flow in `test_native_oft_adapter_send()` so this PR just removes the internal function testing.

File: packages/oft-evm/test/OFT.t.sol
Comment (ravinagill15) [reply packages/oft-evm/test/OFT.t.sol:457]: I decided to make a separate PR for this to keep things cleaner since this PR has already been sent for audit as is

https://github.com/LayerZero-Labs/devtools/pull/880

File: packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol
Comment (ryandgoulding) [reply packages/oft-evm/test/mocks/NativeOFTAdapterMock.sol:26]: Just makes it easier to encapsulated functions in a more controlled manner.  Since the contracts are immutable, it is generally pretty safe as any implementation change is likely going to go through re-audit, redeployment, etc.  But, it is fine to remove them if they are tested elsewhere.  In general, agreed, it is a shortcut to testing the API as it will actually be used.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:8]: Should this be removed now?

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:14]: NIT: "which" needs to be done very carefully.

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ryandgoulding) [inline packages/oft-evm/contracts/NativeOFTAdapter.sol:127]: Why name the return parameter if you are not assigning it?

**Review by ryandgoulding — DISMISSED** [review]

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:8]: Great eye!

File: packages/oft-evm/contracts/NativeOFTAdapter.sol
Comment (ravinagill15) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:127]: That's a great point that I didn't think of! I will keep in mind that return variables should only be named if being assigned. Considering this has already been audited, I am hesitant to change any code even though I removed the return parameter name locally and the tests still passed

Please let me know if you strongly recommend this change, if so I will look into whether the auditors are okay with this minor change 🫡

Comment (ryandgoulding) [reply packages/oft-evm/contracts/NativeOFTAdapter.sol:126]: No it is fine, just wanted to make sure I wasn't overlooking anything.

**Review by janjakubnanista — APPROVED** [review]

**Pull request merged**
